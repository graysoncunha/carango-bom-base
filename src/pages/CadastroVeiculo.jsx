import React, { useEffect, useState } from 'react'
import { useHistory, useParams } from 'react-router'
import { Button, TextField, MenuItem } from '@material-ui/core'
import useErros from '../hooks/useErros'
import VeiculoService from '../services/VeiculoService'
import MarcaService from '../services/MarcaService'

function CadastroVeiculo() {
  const [veiculo, setVeiculo] = useState({
    marca: '',
    modelo: '',
    ano: '',
    valor: '',
  })
  const [marcas, setMarcas] = useState([])

  const history = useHistory()
  const { id } = useParams()

  const validacoes = {
    marca: (dado) => {
      if (dado) {
        return { valido: true }
      } else {
        return { valido: false, texto: 'Marca é obrigatória.' }
      }
    },
    modelo: (dado) => {
      if (!dado || dado.length < 2) {
        return {
          valido: false,
          texto: 'Modelo deve ter ao menos 2 caracteres.',
        }
      } else {
        return { valido: true }
      }
    },
    ano: (dado) => {
      if (!dado || dado.length < 4) {
        return { valido: false, texto: 'Ano deve ter ao menos 4 caracteres.' }
      } else {
        return { valido: true }
      }
    },
    valor: (dado) => {
      if (dado) {
        return { valido: true }
      } else {
        return { valido: false, texto: 'Valor é obrigatório.' }
      }
    },
  }

  const [erros, validarCampos, possoEnviar] = useErros(validacoes)

  useEffect(async () => {
    const marcas = await MarcaService.listar()
    setMarcas(marcas)
  }, [])

  useEffect(async () => {
    if (id) {
      const veiculo = await VeiculoService.consultar(id)

      setVeiculo({
        marca: veiculo.marca.id,
        modelo: veiculo.modelo,
        ano: veiculo.ano,
        valor: veiculo.valor,
      })
    }
  }, [id])

  function cancelar() {
    history.goBack()
  }

  return (
    <form
      onSubmit={(event) => {
        event.preventDefault()

        if (possoEnviar()) {
          if (id) {
            VeiculoService.alterar({ id, ...veiculo }).then((res) => {
              history.goBack()
            })
          } else {
            VeiculoService.cadastrar({ ...veiculo }).then((res) => {
              setVeiculo({
                marca: '',
                modelo: '',
                ano: '',
                valor: '',
              })
              history.goBack()
            })
          }
        }
      }}
    >
      <TextField
        value={veiculo.marca}
        onChange={(evt) =>
          setVeiculo((prevState) => ({
            ...prevState,
            marca: evt.target.value,
          }))
        }
        helperText={erros.marca.texto}
        error={!erros.marca.valido}
        name="marca"
        id="marca"
        label="Marca"
        select
        variant="outlined"
        fullWidth
        required
        margin="normal"
      >
        {marcas &&
          marcas.map(({ id, nome }) => (
            <MenuItem key={id} value={id}>
              {nome}
            </MenuItem>
          ))}
      </TextField>

      <TextField
        value={veiculo.modelo}
        onChange={(evt) =>
          setVeiculo((prevState) => ({
            ...prevState,
            modelo: evt.target.value,
          }))
        }
        onBlur={validarCampos}
        helperText={erros.modelo.texto}
        error={!erros.modelo.valido}
        name="modelo"
        id="modelo"
        label="Modelo"
        type="text"
        variant="outlined"
        fullWidth
        required
        margin="normal"
      />

      <TextField
        value={veiculo.ano}
        onChange={(evt) =>
          setVeiculo((prevState) => ({
            ...prevState,
            ano: evt.target.value,
          }))
        }
        onBlur={validarCampos}
        helperText={erros.ano.texto}
        error={!erros.ano.valido}
        name="ano"
        id="ano"
        label="Ano"
        type="text"
        variant="outlined"
        fullWidth
        required
        margin="normal"
      />

      <TextField
        value={veiculo.valor}
        onChange={(evt) =>
          setVeiculo((prevState) => ({
            ...prevState,
            valor: evt.target.value,
          }))
        }
        onBlur={validarCampos}
        helperText={erros.valor.texto}
        error={!erros.valor.valido}
        name="valor"
        id="valor"
        label="Valor"
        type="text"
        variant="outlined"
        fullWidth
        required
        margin="normal"
      />

      <Button variant="contained" color="primary" type="submit" disabled={!possoEnviar()}>
        {id ? 'Alterar' : 'Cadastrar'}
      </Button>

      <Button variant="contained" color="secondary" onClick={cancelar}>
        Cancelar
      </Button>
    </form>
  )
}

export default CadastroVeiculo
