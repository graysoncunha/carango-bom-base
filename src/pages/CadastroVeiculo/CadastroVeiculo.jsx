import React, { useEffect, useState } from 'react'
import { useHistory, useParams } from 'react-router'
import { Button, TextField, MenuItem, makeStyles } from '@material-ui/core'
import Alert from '@material-ui/lab/Alert'
import useErros from '../../hooks/useErros'
import VeiculoService from '../../services/VeiculoService'
import MarcaService from '../../services/MarcaService'

const useStyles = makeStyles((theme) => ({
  actions: {
    top: '10px',
    marginRight: '10px',
  },
  alert: {
    width: '100%',
    marginBottom: theme.spacing(2),
  },
}))

function CadastroVeiculo() {
  const [{ status, error }, setStatus] = useState({ status: 'idle', error: null })
  const [veiculo, setVeiculo] = useState({
    marcaId: '',
    modelo: '',
    ano: '',
    valor: '',
  })
  const [marcas, setMarcas] = useState([])

  const classes = useStyles()
  const history = useHistory()
  const { id } = useParams()

  const validacoes = {
    marca: (dado) => {
      if (dado) {
        return { valido: true }
      } else {
        return { valido: false, texto: 'Marca é obrigatória.' }
      }
    },
    modelo: (dado) => {
      if (!dado || dado.length < 2) {
        return {
          valido: false,
          texto: 'Modelo deve ter ao menos 2 caracteres.',
        }
      } else {
        return { valido: true }
      }
    },
    ano: (dado) => {
      if (!dado || dado.length < 4) {
        return { valido: false, texto: 'Ano deve ter ao menos 4 caracteres.' }
      } else {
        return { valido: true }
      }
    },
    valor: (dado) => {
      if (dado) {
        return { valido: true }
      } else {
        return { valido: false, texto: 'Valor é obrigatório.' }
      }
    },
  }

  const [erros, validarCampos, possoEnviar] = useErros(validacoes)

  useEffect(async () => {
    const marcasLista = await MarcaService.listar()
    setMarcas(marcasLista)
  }, [])

  useEffect(async () => {
    if (id) {
      const veiculoSelecionado = await VeiculoService.consultar(id)

      setVeiculo({
        marcaId: veiculoSelecionado.marca.id,
        modelo: veiculoSelecionado.modelo,
        ano: veiculoSelecionado.ano,
        valor: veiculoSelecionado.valor,
      })
    }
  }, [id])

  function cancelar() {
    history.goBack()
  }

  async function cadastrar() {
    try {
      setStatus({ status: 'loading' })
      await VeiculoService.cadastrar({ ...veiculo })
      setVeiculo({
        marcaId: '',
        modelo: '',
        ano: '',
        valor: '',
      })
      history.goBack()
      setStatus({ status: 'fulfilled' })
    } catch (e) {
      setStatus({ status: 'rejected', error: e.message })
    }
  }

  return (
    <div>
      {status === 'rejected' ? (
        <Alert severity="error" className={classes.alert}>
          {error}
        </Alert>
      ) : null}
      <form
        onSubmit={(event) => {
          event.preventDefault()

          if (possoEnviar()) {
            if (id) {
              VeiculoService.alterar({ id, ...veiculo }).then(() => {
                history.goBack()
              })
            } else {
              cadastrar()
            }
          }
        }}
      >
        <TextField
          value={veiculo.marcaId}
          onChange={(evt) =>
            setVeiculo((prevState) => ({
              ...prevState,
              marcaId: evt.target.value,
            }))
          }
          helperText={erros.marca.texto}
          error={!erros.marca.valido}
          name="marca"
          id="marca"
          label="Marca"
          select
          variant="outlined"
          fullWidth
          required
          margin="normal"
        >
          {marcas &&
            marcas.map((marca) => (
              <MenuItem key={marca.id} value={marca.id}>
                {marca.nome}
              </MenuItem>
            ))}
        </TextField>

        <TextField
          value={veiculo.modelo}
          onChange={(evt) =>
            setVeiculo((prevState) => ({
              ...prevState,
              modelo: evt.target.value,
            }))
          }
          onBlur={validarCampos}
          helperText={erros.modelo.texto}
          error={!erros.modelo.valido}
          name="modelo"
          id="modelo"
          label="Modelo"
          type="text"
          variant="outlined"
          fullWidth
          required
          margin="normal"
        />

        <TextField
          value={veiculo.ano}
          onChange={(evt) =>
            setVeiculo((prevState) => ({
              ...prevState,
              ano: evt.target.value,
            }))
          }
          onBlur={validarCampos}
          helperText={erros.ano.texto}
          error={!erros.ano.valido}
          name="ano"
          id="ano"
          label="Ano"
          type="text"
          variant="outlined"
          fullWidth
          required
          margin="normal"
        />

        <TextField
          value={veiculo.valor}
          onChange={(evt) =>
            setVeiculo((prevState) => ({
              ...prevState,
              valor: evt.target.value,
            }))
          }
          onBlur={validarCampos}
          helperText={erros.valor.texto}
          error={!erros.valor.valido}
          name="valor"
          id="valor"
          label="Valor"
          type="text"
          variant="outlined"
          fullWidth
          required
          margin="normal"
        />

        <Button variant="contained" color="primary" type="submit" disabled={!possoEnviar()} className={classes.actions}>
          {id ? 'Alterar' : 'Cadastrar'}
        </Button>

        <Button variant="contained" color="secondary" onClick={cancelar} className={classes.actions}>
          Cancelar
        </Button>
      </form>
    </div>
  )
}

export default CadastroVeiculo
