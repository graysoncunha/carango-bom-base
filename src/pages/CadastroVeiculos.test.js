import React from 'react'
import { render, screen } from '@testing-library/react'
import { MemoryRouter, BrowserRouter } from 'react-router-dom'
import CadastroVeiculo from './CadastroVeiculo'

describe('Componente de cadastro de veículos', () => {
  it('deve carregar a página de cadastro de veículo', () => {
    const { container } = render(<CadastroVeiculo />, { wrapper: MemoryRouter })

    expect(container.firstChild).toMatchSnapshot()
  })

  it('deve carregar a página de alteração de veículo', async () => {
    renderWithRouter(<CadastroVeiculo />, { route: '/alteracao-veiculo/139' })

    expect((await screen.findByText('Alterar')).textContent).toBeInTheDocument()
  })

  // TODO: Ver se a função submit foi chamada com o clique do botão
  //   it('quando o formulário é enviado, chama a função de cadastrar', () => {
  //     const funcaoRealizarTransacao = jest.fn()

  //     render(<Conta saldo={1000} realizarTransacao={funcaoRealizarTransacao} />)

  //     fireEvent.click(screen.getByText('Realizar operação'))

  //     expect(funcaoRealizarTransacao).toHaveBeenCalled()
  //   })
})

const renderWithRouter = (ui, { route = '/' } = {}) => {
  history.pushState({}, 'Test page', route)

  return render(ui, { wrapper: BrowserRouter })
}
