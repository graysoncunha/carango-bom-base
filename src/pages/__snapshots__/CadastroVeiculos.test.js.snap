// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Componente de cadastro de veículos deve carregar a página de cadastro de veículo 1`] = `
<form>
  <div
    class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
      data-shrink="false"
      for="marca"
      id="marca-label"
    >
      Marca
      <span
        aria-hidden="true"
        class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
      >
         
        *
      </span>
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
    >
      <div
        aria-haspopup="listbox"
        aria-labelledby="marca-label marca"
        class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input"
        id="marca"
        role="button"
        tabindex="0"
      >
        <span>
          ​
        </span>
      </div>
      <input
        aria-hidden="true"
        class="MuiSelect-nativeInput"
        name="marca"
        required=""
        tabindex="-1"
        value=""
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
      <fieldset
        aria-hidden="true"
        class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="PrivateNotchedOutline-legendLabelled-3"
        >
          <span>
            Marca
             *
          </span>
        </legend>
      </fieldset>
    </div>
    
  </div>
  <div
    class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
      data-shrink="false"
      for="modelo"
      id="modelo-label"
    >
      Modelo
      <span
        aria-hidden="true"
        class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
      >
         
        *
      </span>
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiOutlinedInput-input"
        id="modelo"
        name="modelo"
        required=""
        type="text"
        value=""
      />
      <fieldset
        aria-hidden="true"
        class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="PrivateNotchedOutline-legendLabelled-3"
        >
          <span>
            Modelo
             *
          </span>
        </legend>
      </fieldset>
    </div>
    
  </div>
  <div
    class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
      data-shrink="false"
      for="ano"
      id="ano-label"
    >
      Ano
      <span
        aria-hidden="true"
        class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
      >
         
        *
      </span>
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiOutlinedInput-input"
        id="ano"
        name="ano"
        required=""
        type="text"
        value=""
      />
      <fieldset
        aria-hidden="true"
        class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="PrivateNotchedOutline-legendLabelled-3"
        >
          <span>
            Ano
             *
          </span>
        </legend>
      </fieldset>
    </div>
    
  </div>
  <div
    class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
      data-shrink="false"
      for="valor"
      id="valor-label"
    >
      Valor
      <span
        aria-hidden="true"
        class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
      >
         
        *
      </span>
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiOutlinedInput-input"
        id="valor"
        name="valor"
        required=""
        type="text"
        value=""
      />
      <fieldset
        aria-hidden="true"
        class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="PrivateNotchedOutline-legendLabelled-3"
        >
          <span>
            Valor
             *
          </span>
        </legend>
      </fieldset>
    </div>
    
  </div>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
    tabindex="0"
    type="submit"
  >
    <span
      class="MuiButton-label"
    >
      Cadastrar
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      Cancelar
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
</form>
`;
